File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/sctokenid.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from "../sandbox";
import * as wasmtypes from "./index";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScTokenIDLength = 38;

export class ScTokenID {
    id: u8[] = wasmtypes.zeroes(ScTokenIDLength);

    public equals(other: ScTokenID): bool {
        return wasmtypes.bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): u8[] {
        return tokenIDToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        // TODO standardize human readable string
        return tokenIDToString(this);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function tokenIDDecode(dec: wasmtypes.WasmDecoder): ScTokenID {
    return tokenIDFromBytesUnchecked(dec.fixedBytes(ScTokenIDLength));
}

export function tokenIDEncode(enc: wasmtypes.WasmEncoder, value: ScTokenID): void {
    enc.fixedBytes(value.id, ScTokenIDLength);
}

export function tokenIDFromBytes(buf: u8[]): ScTokenID {
    if (buf.length == 0) {
        return new ScTokenID();
    }
    if (buf.length != ScTokenIDLength) {
        panic("invalid TokenID length");
    }
    return tokenIDFromBytesUnchecked(buf);
}

export function tokenIDToBytes(value: ScTokenID): u8[] {
    return value.id;
}

export function tokenIDToString(value: ScTokenID): string {
    // TODO standardize human readable string
    return wasmtypes.base58Encode(value.id);
}

function tokenIDFromBytesUnchecked(buf: u8[]): ScTokenID {
    let o = new ScTokenID();
    o.id = buf.slice(0);
    return o;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableTokenID {
    proxy: wasmtypes.Proxy;

    constructor(proxy: wasmtypes.Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return tokenIDToString(this.value());
    }

    value(): ScTokenID {
        return tokenIDFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableTokenID extends ScImmutableTokenID {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScTokenID): void {
        this.proxy.set(tokenIDToBytes(value));
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/sctokenid.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//nolint:dupl
package wasmtypes

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const ScTokenIDLength = 38

type ScTokenID struct {
	id [ScTokenIDLength]byte
}

func (o ScTokenID) Bytes() []byte {
	return TokenIDToBytes(o)
}

func (o ScTokenID) String() string {
	return TokenIDToString(o)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func TokenIDDecode(dec *WasmDecoder) ScTokenID {
	return tokenIDFromBytesUnchecked(dec.FixedBytes(ScTokenIDLength))
}

func TokenIDEncode(enc *WasmEncoder, value ScTokenID) {
	enc.FixedBytes(value.id[:], ScTokenIDLength)
}

func TokenIDFromBytes(buf []byte) ScTokenID {
	if len(buf) == 0 {
		return ScTokenID{}
	}
	if len(buf) != ScTokenIDLength {
		panic("invalid TokenID length")
	}
	return tokenIDFromBytesUnchecked(buf)
}

func TokenIDToBytes(value ScTokenID) []byte {
	return value.id[:]
}

func TokenIDFromString(value string) ScTokenID {
	return TokenIDFromBytes(HexDecode(value[2:]))
}

func TokenIDToString(value ScTokenID) string {
	return "0x" + HexEncode(TokenIDToBytes(value))
}

func tokenIDFromBytesUnchecked(buf []byte) ScTokenID {
	o := ScTokenID{}
	copy(o.id[:], buf)
	return o
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableTokenID struct {
	proxy Proxy
}

func NewScImmutableTokenID(proxy Proxy) ScImmutableTokenID {
	return ScImmutableTokenID{proxy: proxy}
}

func (o ScImmutableTokenID) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableTokenID) String() string {
	return TokenIDToString(o.Value())
}

func (o ScImmutableTokenID) Value() ScTokenID {
	return TokenIDFromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableTokenID struct {
	ScImmutableTokenID
}

func NewScMutableTokenID(proxy Proxy) ScMutableTokenID {
	return ScMutableTokenID{ScImmutableTokenID{proxy: proxy}}
}

func (o ScMutableTokenID) Delete() {
	o.proxy.Delete()
}

func (o ScMutableTokenID) SetValue(value ScTokenID) {
	o.proxy.Set(TokenIDToBytes(value))
}
